### PR TITLE
Merge retry failed push fixes #201

### LINKED ISSUE #201: Retry mechanism for failed push [$15 awarded]
If the push failed because some changes from remote were not fetched, we should try to fetch, merge them and in the end, retry the push.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26277156-retry-mechanism-for-failed-push?utm_campaign=plugin&utm_content=tracker%2F7081451&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F7081451&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by krodyrobi**
@PressLabs/presslabs CR

If there is anything else regarding possible tests. I'm open to ideas.

### CODE REVIEW

File: gitfs/merges/accept_mine.py [hunk 1]
```diff
@@ -85,16 +85,23 @@ def __call__(self, local_branch, remote_branch, upstream):
...
+        for branch, ref in refs:
+            log.debug("AcceptMine: Delete %s" % branch)
+            ref = self.repository.lookup_reference(ref)
+            ref.delete()
+
+    def __call__(self, local_branch, remote_branch, upstream):
+        try:
+            self._merge(local_branch, remote_branch, upstream)
+        finally:
+            self._merge_clean_up(local_branch)
```
Comment (krodyrobi) [inline gitfs/merges/accept_mine.py]: If an exception occurred anywhere in the try or the finally block the exception will get bubbled upwards.

**Comment by krodyrobi** [comment]
@PressLabs/presslabs  CR?

File: gitfs/merges/accept_mine.py [hunk 2]
```diff
@@ -33,11 +33,15 @@ def _create_remote_copy(self, branch_name, upstream, new_branch):
         return local
 
     def _create_local_copy(self, branch_name, new_branch):
-        old_branch = self.repository.lookup_branch(branch_name,
-                                                   pygit2.GIT_BRANCH_LOCAL)
-        return old_branch.rename(new_branch, True)
+        branch = self.repository.lookup_branch(branch_name,
```
Comment (calind) [inline gitfs/merges/accept_mine.py]: `branch` is assigned but not used.

Comment (krodyrobi) [inline gitfs/merges/accept_mine.py]: The branch variable is used select the split option in the viewer on this page to see it better :D the function continues after the removal of the call function.

Comment (krodyrobi) [inline gitfs/merges/accept_mine.py]: The branch variable is used select the split option in the viewer on this page to see it better :D the function continues after the removal of the call function.

File: gitfs/worker/sync.py [hunk 3]
```diff
@@ -87,7 +87,9 @@ def on_idle(self):
                 self.commit(self.commits)
                 self.commits = []
             log.debug("Start syncing")
-            self.sync()
+            while not self.sync():
```
Comment (rciorba) [inline gitfs/worker/sync.py]: should this have some sort of back-off mechanism? sleep for a while maybe? right now this has the potential to become a cpu-hogging busy loop

Comment (krodyrobi) [inline gitfs/worker/sync.py]: The only thing to note here is that I believe that if the sync worker stalls new local commits won't be added to the queue for further pushing for that period.

But it shouldn't be the case. The max wait time is 33s.

**Comment by krodyrobi** [comment]
@vtemian currently if no internet connection is available it will continuously try to push without committing changes locally. Is this the way it should work? 

Or it should stop trying to merge the remote after some number of tries and resume the local commit phase then try again with the remote merge?

I find the second option more natural as this way it would save local progress at least x amount of time.

**Comment by vtemian** [comment]
Switching the filesystem to read-only if a push failed it's the right thing to do. In this way you will be POSIX complaint, when a write occured and the write system call succeeded, you're information needs to be on the upstream. If this it's not possible, wait until it's possible and in case of conflict, apply `always accept mine` strategy.

Almost the same scenario in case the commit queue it's not already fully consumed. You're content won't be online if a failure was accoutered.

File: gitfs/merges/accept_mine.py [hunk 4]
```diff
@@ -33,11 +33,15 @@ def _create_remote_copy(self, branch_name, upstream, new_branch):
...
+                                               pygit2.GIT_BRANCH_LOCAL)
 
-    def __call__(self, local_branch, remote_branch, upstream):
+        branch_commit = branch.get_object()
+        local = self.repository.create_branch(new_branch, branch_commit)
```
Comment (vtemian) [inline gitfs/merges/accept_mine.py]: You can just return `self.repository.create_branch(new_branch, branch_commit)`

File: gitfs/merges/accept_mine.py [hunk 5]
```diff
@@ -33,11 +33,15 @@ def _create_remote_copy(self, branch_name, upstream, new_branch):
...
+
+        return local
+
+    def _merge(self, local_branch, remote_branch, upstream):
```
Comment (vtemian) [inline gitfs/merges/accept_mine.py]: Rename it to `merge`

File: gitfs/merges/accept_mine.py [hunk 6]
```diff
@@ -85,16 +89,26 @@ def __call__(self, local_branch, remote_branch, upstream):
         self.repository.create_reference("refs/heads/%s" % local_branch,
                                          ref.target,
                                          force=True)
+
+    def _merge_clean_up(self, local_branch):
```
Comment (vtemian) [inline gitfs/merges/accept_mine.py]: Rename it to `clean_up`

**Comment by krodyrobi** [comment]
@vtemian can't seem to find a way to set the fs as readonly? Can this be done with the current impl?
Any hints to reduce the codebase would be great.

**Comment by vtemian** [comment]
This is the current behavior. It's already implemented, that's way if a push fails it will try to push without committing local changes (which in theory you shouldn't have, because the filesystem is in read-only)

**Comment by zalmanu** [comment]
@krodyrobi can you please take a look at the suggested changes?

**Comment by krodyrobi** [comment]
@PressLabs/presslabs CR

**Pull request merged**